File: CheckAuto2/app/src/main/java/com/tecsup/checkauto/service/SupabaseService.kt
package com.tecsup.checkauto.service

import com.tecsup.checkauto.config.SupabaseConfig
import com.tecsup.checkauto.model.Anuncio
import com.tecsup.checkauto.model.Vehiculo
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.query.Columns
import io.github.jan.supabase.postgrest.query.Order
import io.github.jan.supabase.storage.storage
import kotlinx.serialization.Serializable

// ========================================
// MODELOS DE DATOS PARA SUPABASE
// ========================================

@Serializable
data class AnuncioSupabase(
    val id_anuncio: Int? = null,
    val id_usuario: String,
    val titulo: String? = null,
    val modelo: String,
    val anio: Int,
    val kilometraje: Int,
    val precio: Double,
    val descripcion: String,
    val email_contacto: String? = null,
    val telefono_contacto: String? = null,
    val tipo_vehiculo: String? = null,
    val fecha_creacion: String? = null,
    val fecha_actualizacion: String? = null,
    val activo: Boolean = true
)

@Serializable
data class ImagenSupabase(
    val id_imagen: Int? = null,
    val id_anuncio: Int,
    val url_imagen: String,
    val nombre_archivo: String? = null,
    val tipo_archivo: String? = null,
    val tamano_archivo: Long? = null,
    val fecha_subida: String? = null,
    val orden: Int = 0
)

@Serializable
data class NotificacionSupabase(
    val id_notificacion: Int? = null,
    val id_usuario: Int? = null,
    val id_vendedor: String? = null,
    val id_comprador: String? = null,
    val nombre_comprador: String? = null,
    val email_comprador: String? = null,
    val id_anuncio: Int? = null,
    val titulo: String? = null,
    val mensaje: String? = null,
    val leido: Boolean = false,
    val leida: Boolean = false,
    val fecha_creacion: String? = null,
    val metadata: String? = null,
    val tipo: String = "interes"
)

@Serializable
data class VehiculoSupabase(
    val id_vehiculo: Int? = null,
    val placa: String? = null,
    val descripcion_api: String? = null,
    val marca: String? = null,
    val modelo: String? = null,
    val anio_registro_api: String? = null,
    val vin: String? = null,
    val uso: String? = null,
    val propietario: String? = null,
    val delivery_point: String? = null,
    val fecha_registro_api: String? = null,
    val image_url_api: String? = null,
    val datos_api: String? = null,
    val fecha_actualizacion_api: String? = null,
    val tipo_vehiculo: String? = null
)

@Serializable
data class HistorialBusquedaSupabase(
    val id_historial: Int? = null,
    val id_usuario: String? = null,
    val placa_consultada: String? = null,
    val fecha_consulta: String? = null,
    val resultado_api: String? = null
)

@Serializable
data class CategoriaVehiculoSupabase(
    val id_categoria: Int? = null,
    val nombre: String,
    val codigo: String? = null,
    val descripcion: String? = null,
    val activo: Boolean = true,
    val fecha_creacion: String? = null
)

// ========================================
// SERVICIO DE SUPABASE
// ========================================

object SupabaseService {
    private val client = SupabaseConfig.client
    
    // ========================================
    // ANUNCIOS
    // ========================================
    
    suspend fun getAnuncios(): List<AnuncioSupabase> {
        return client.from("anuncios")
            .select {
                filter {
                    eq("activo", true)
                }
                order("fecha_creacion", Order.DESCENDING)
            }
            .decodeList<AnuncioSupabase>()
    }
    
    suspend fun getAnuncioById(id: Int): AnuncioSupabase? {
        return client.from("anuncios")
            .select {
                filter {
                    eq("id_anuncio", id)
                }
            }
            .decodeSingleOrNull<AnuncioSupabase>()
    }
    
    suspend fun getAnunciosByUserId(userId: String): List<AnuncioSupabase> {
        return client.from("anuncios")
            .select {
                filter {
                    eq("id_usuario", userId)
                }
                order("fecha_creacion", Order.DESCENDING)
            }
            .decodeList<AnuncioSupabase>()
    }
    
    suspend fun createAnuncio(anuncio: AnuncioSupabase): AnuncioSupabase {
        return client.from("anuncios")
            .insert(anuncio) {
                select()
            }
            .decodeSingle<AnuncioSupabase>()
    }
    
    suspend fun updateAnuncio(id: Int, anuncio: AnuncioSupabase): AnuncioSupabase {
        return client.from("anuncios")
            .update(anuncio) {
                filter {
                    eq("id_anuncio", id)
                }
                select()
            }
            .decodeSingle<AnuncioSupabase>()
    }
    
    suspend fun deleteAnuncio(id: Int) {
        client.from("anuncios")
            .update(mapOf("activo" to false)) {
                filter {
                    eq("id_anuncio", id)
                }
            }
    }
    
    // ========================================
    // IMÁGENES
    // ========================================
    
    suspend fun getImagenesByAnuncioId(anuncioId: Int): List<ImagenSupabase> {
        return client.from("imagenes")
            .select {
                filter {
                    eq("id_anuncio", anuncioId)
                }
                order("orden", Order.ASCENDING)
            }
            .decodeList<ImagenSupabase>()
    }
    
    suspend fun addImagen(imagen: ImagenSupabase): ImagenSupabase {
        return client.from("imagenes")
            .insert(imagen) {
                select()
            }
            .decodeSingle<ImagenSupabase>()
    }
    
    suspend fun deleteImagen(id: Int) {
        client.from("imagenes")
            .delete {
                filter {
                    eq("id_imagen", id)
                }
            }
    }
    
    // ========================================
    // NOTIFICACIONES
    // ========================================
    
    suspend fun getNotificacionesByVendedor(vendedorId: String): List<NotificacionSupabase> {
        return client.from("notificaciones")
            .select {
                filter {
                    eq("id_vendedor", vendedorId)
                }
                order("fecha_creacion", Order.DESCENDING)
            }
            .decodeList<NotificacionSupabase>()
    }
    
    suspend fun getUnreadNotificaciones(vendedorId: String): List<NotificacionSupabase> {
        return client.from("notificaciones")
            .select {
                filter {
                    eq("id_vendedor", vendedorId)
                    eq("leido", false)
                }
                order("fecha_creacion", Order.DESCENDING)
            }
            .decodeList<NotificacionSupabase>()
    }
    
    suspend fun createNotificacion(notificacion: NotificacionSupabase): NotificacionSupabase {
        return client.from("notificaciones")
            .insert(notificacion) {
                select()
            }
            .decodeSingle<NotificacionSupabase>()
    }
    
    suspend fun markNotificacionAsRead(id: Int) {
        client.from("notificaciones")
            .update(mapOf("leido" to true, "leida" to true)) {
                filter {
                    eq("id_notificacion", id)
                }
            }
    }
    
    suspend fun markAllNotificacionesAsRead(vendedorId: String) {
        client.from("notificaciones")
            .update(mapOf("leido" to true, "leida" to true)) {
                filter {
                    eq("id_vendedor", vendedorId)
                    eq("leido", false)
                }
            }
    }
    
    // ========================================
    // VEHÍCULOS
    // ========================================
    
    suspend fun getVehiculoByPlaca(placa: String): VehiculoSupabase? {
        return client.from("vehiculos")
            .select {
                filter {
                    eq("placa", placa)
                }
            }
            .decodeSingleOrNull<VehiculoSupabase>()
    }
    
    suspend fun upsertVehiculo(vehiculo: VehiculoSupabase): VehiculoSupabase {
        return client.from("vehiculos")
            .upsert(vehiculo) {
                select()
            }
            .decodeSingle<VehiculoSupabase>()
    }
    
    // ========================================
    // HISTORIAL DE BÚSQUEDA
    // ========================================
    
    suspend fun getHistorialByUserId(userId: String, limit: Int = 10): List<HistorialBusquedaSupabase> {
        return client.from("historial_busqueda")
            .select {
                filter {
                    eq("id_usuario", userId)
                }
                order("fecha_consulta", Order.DESCENDING)
                limit(limit.toLong())
            }
            .decodeList<HistorialBusquedaSupabase>()
    }
    
    suspend fun addHistorial(historial: HistorialBusquedaSupabase): HistorialBusquedaSupabase {
        return client.from("historial_busqueda")
            .insert(historial) {
                select()
            }
            .decodeSingle<HistorialBusquedaSupabase>()
    }
    
    // ========================================
    // STORAGE - IMÁGENES
    // ========================================
    
    /**
     * Subir imagen a Supabase Storage
     * @param bucket Nombre del bucket (ej: "anuncios")
     * @param path Ruta donde guardar (ej: "anuncio-123/imagen1.jpg")
     * @param data Bytes de la imagen
     * @return URL pública de la imagen
     */
    suspend fun uploadImage(
        bucket: String,
        path: String,
        data: ByteArray
    ): String {
        val storage = client.storage.from(bucket)
        storage.upload(path, data) {
            upsert = true
        }
        return storage.publicUrl(path)
    }
    
    /**
     * Eliminar imagen de Supabase Storage
     */
    suspend fun deleteImage(bucket: String, path: String) {
        client.storage.from(bucket).delete(path)
    }
    
    /**
     * Obtener URL pública de una imagen
     */
    suspend fun getImageUrl(bucket: String, path: String): String {
        return client.storage.from(bucket).publicUrl(path)
    }
    
    // ========================================
    // CATEGORÍAS DE VEHÍCULOS
    // ========================================
    
    /**
     * Obtener todas las categorías de vehículos activas
     */
    suspend fun getCategoriasVehiculos(): List<CategoriaVehiculoSupabase> {
        return client.from("categorias_vehiculos")
            .select {
                filter {
                    eq("activo", true)
                }
                order("nombre", Order.ASCENDING)
            }
            .decodeList<CategoriaVehiculoSupabase>()
    }
    
    /**
     * Obtener categoría por ID
     */
    suspend fun getCategoriaById(id: Int): CategoriaVehiculoSupabase? {
        return client.from("categorias_vehiculos")
            .select {
                filter {
                    eq("id_categoria", id)
                }
            }
            .decodeSingleOrNull<CategoriaVehiculoSupabase>()
    }
    
    /**
     * Obtener categoría por código
     */
    suspend fun getCategoriaByCodigo(codigo: String): CategoriaVehiculoSupabase? {
        return client.from("categorias_vehiculos")
            .select {
                filter {
                    eq("codigo", codigo)
                    eq("activo", true)
                }
            }
            .decodeSingleOrNull<CategoriaVehiculoSupabase>()
    }
}

File: react-front/src/services/supabaseService.js
import { supabase } from '../config/supabase';

// ========================================
// SERVICIO DE ANUNCIOS
// ========================================
export const anunciosService = {
  // Obtener todos los anuncios activos
  async getAll(filters = {}) {
    let query = supabase
      .from('anuncios')
      .select(`
        *,
        imagenes (*)
      `)
      .eq('activo', true)
      .order('fecha_creacion', { ascending: false });

    // Aplicar filtros opcionales
    if (filters.tipo_vehiculo) {
      query = query.eq('tipo_vehiculo', filters.tipo_vehiculo);
    }
    if (filters.precio_min) {
      query = query.gte('precio', filters.precio_min);
    }
    if (filters.precio_max) {
      query = query.lte('precio', filters.precio_max);
    }
    if (filters.anio_min) {
      query = query.gte('anio', filters.anio_min);
    }
    if (filters.limit) {
      query = query.limit(filters.limit);
    }

    const { data, error } = await query;
    if (error) throw error;
    return data;
  },

  // Obtener un anuncio por ID
  async getById(id) {
    const { data, error } = await supabase
      .from('anuncios')
      .select(`
        *,
        imagenes (*)
      `)
      .eq('id_anuncio', id)
      .single();

    if (error) throw error;
    return data;
  },

  // Obtener anuncios de un usuario específico
  async getByUserId(userId) {
    const { data, error } = await supabase
      .from('anuncios')
      .select(`
        *,
        imagenes (*)
      `)
      .eq('id_usuario', userId)
      .order('fecha_creacion', { ascending: false });

    if (error) throw error;
    return data;
  },

  // Crear nuevo anuncio
  async create(anuncioData) {
    const { data, error } = await supabase
      .from('anuncios')
      .insert([anuncioData])
      .select()
      .single();

    if (error) throw error;
    return data;
  },

  // Actualizar anuncio
  async update(id, anuncioData) {
    const { data, error } = await supabase
      .from('anuncios')
      .update({
        ...anuncioData,
        fecha_actualizacion: new Date().toISOString()
      })
      .eq('id_anuncio', id)
      .select()
      .single();

    if (error) throw error;
    return data;
  },

  // Eliminar anuncio (soft delete)
  async delete(id) {
    const { data, error } = await supabase
      .from('anuncios')
      .update({ activo: false })
      .eq('id_anuncio', id)
      .select()
      .single();

    if (error) throw error;
    return data;
  },

  // Eliminar anuncio permanentemente
  async hardDelete(id) {
    const { error } = await supabase
      .from('anuncios')
      .delete()
      .eq('id_anuncio', id);

    if (error) throw error;
    return true;
  }
};

// ========================================
// SERVICIO DE IMÁGENES
// ========================================
export const imagenesService = {
  // Agregar imagen a un anuncio
  async add(imagenData) {
    const { data, error } = await supabase
      .from('imagenes')
      .insert([imagenData])
      .select()
      .single();

    if (error) throw error;
    return data;
  },

  // Obtener imágenes de un anuncio
  async getByAnuncioId(anuncioId) {
    const { data, error } = await supabase
      .from('imagenes')
      .select('*')
      .eq('id_anuncio', anuncioId)
      .order('orden', { ascending: true });

    if (error) throw error;
    return data;
  },

  // Eliminar imagen
  async delete(id) {
    const { error } = await supabase
      .from('imagenes')
      .delete()
      .eq('id_imagen', id);

    if (error) throw error;
    return true;
  },

  // Subir imagen al storage de Supabase
  async uploadToStorage(file, fileName) {
    const { data, error } = await supabase.storage
      .from('anuncios-images')
      .upload(fileName, file, {
        cacheControl: '3600',
        upsert: false
      });

    if (error) throw error;
    
    // Obtener URL pública
    const { data: urlData } = supabase.storage
      .from('anuncios-images')
      .getPublicUrl(fileName);

    return urlData.publicUrl;
  }
};

// ========================================
// SERVICIO DE VEHÍCULOS
// ========================================
export const vehiculosService = {
  // Obtener vehículo por placa
  async getByPlaca(placa) {
    const { data, error } = await supabase
      .from('vehiculos')
      .select('*')
      .eq('placa', placa)
      .single();

    if (error && error.code !== 'PGRST116') throw error; // PGRST116 = no rows
    return data;
  },

  // Crear o actualizar vehículo
  async upsert(vehiculoData) {
    const { data, error } = await supabase
      .from('vehiculos')
      .upsert([vehiculoData], { onConflict: 'placa' })
      .select()
      .single();

    if (error) throw error;
    return data;
  },

  // Obtener todos los vehículos
  async getAll() {
    const { data, error } = await supabase
      .from('vehiculos')
      .select('*')
      .order('fecha_actualizacion_api', { ascending: false });

    if (error) throw error;
    return data;
  }
};

// ========================================
// SERVICIO DE NOTIFICACIONES
// ========================================
export const notificacionesService = {
  // Obtener notificaciones de un usuario (como vendedor)
  async getByVendedor(vendedorId) {
    const { data, error } = await supabase
      .from('notificaciones')
      .select(`
        *,
        anuncios:id_anuncio (titulo, modelo, precio)
      `)
      .eq('id_vendedor', vendedorId)
      .order('fecha_creacion', { ascending: false });

    if (error) throw error;
    return data;
  },

  // Obtener notificaciones no leídas
  async getUnread(vendedorId) {
    const { data, error } = await supabase
      .from('notificaciones')
      .select('*')
      .eq('id_vendedor', vendedorId)
      .eq('leido', false)
      .order('fecha_creacion', { ascending: false });

    if (error) throw error;
    return data;
  },

  // Contar notificaciones no leídas
  async countUnread(vendedorId) {
    const { count, error } = await supabase
      .from('notificaciones')
      .select('*', { count: 'exact', head: true })
      .eq('id_vendedor', vendedorId)
      .eq('leido', false);

    if (error) throw error;
    return count;
  },

  // Crear notificación (cuando alguien muestra interés)
  async create(notificacionData) {
    const { data, error } = await supabase
      .from('notificaciones')
      .insert([{
        ...notificacionData,
        tipo: notificacionData.tipo || 'interes'
      }])
      .select()
      .single();

    if (error) throw error;
    return data;
  },

  // Marcar como leída
  async markAsRead(id) {
    const { data, error } = await supabase
      .from('notificaciones')
      .update({ leido: true, leida: true })
      .eq('id_notificacion', id)
      .select()
      .single();

    if (error) throw error;
    return data;
  },

  // Marcar todas como leídas
  async markAllAsRead(vendedorId) {
    const { data, error } = await supabase
      .from('notificaciones')
      .update({ leido: true, leida: true })
      .eq('id_vendedor', vendedorId)
      .eq('leido', false)
      .select();

    if (error) throw error;
    return data;
  },

  // Suscribirse a notificaciones en tiempo real
  subscribeToNewNotifications(vendedorId, callback) {
    return supabase
      .channel('notificaciones-channel')
      .on(
        'postgres_changes',
        {
          event: 'INSERT',
          schema: 'public',
          table: 'notificaciones',
          filter: `id_vendedor=eq.${vendedorId}`
        },
        (payload) => callback(payload.new)
      )
      .subscribe();
  }
};

// ========================================
// SERVICIO DE HISTORIAL DE BÚSQUEDA
// ========================================
export const historialService = {
  // Obtener historial de un usuario
  async getByUserId(userId, limit = 10) {
    const { data, error } = await supabase
      .from('historial_busqueda')
      .select('*')
      .eq('id_usuario', userId)
      .order('fecha_consulta', { ascending: false })
      .limit(limit);

    if (error) throw error;
    return data;
  },

  // Agregar búsqueda al historial
  async add(historialData) {
    const { data, error } = await supabase
      .from('historial_busqueda')
      .insert([historialData])
      .select()
      .single();

    if (error) throw error;
    return data;
  }
};

// Exportar todos los servicios
export default {
  anuncios: anunciosService,
  imagenes: imagenesService,
  vehiculos: vehiculosService,
  notificaciones: notificacionesService,
  historial: historialService
};

